rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,183 @@
+```ln
+let conditionalTable = newarr(2);
+const condition1Boolean = val.isSome();
+const condition1ThenFn = fn(): U { return val.getMaybe().toString(); };
+condfn(conditionalTable, condition1Boolean, condition1ThenFn);
+const condition2Boolean = true;
+const condition2ThenFn = fn(): U { return 'none'; };
+condfn(conditionalTable, condition2Boolean, condition2ThenFn);
+const conditionTailFn = fn(): U {};
+const conditionResult: Maybe<U> = evalcond(conditionalTable, conditionTailFn);
+const conditionResultUnwrapped: U = conditionResult.getMaybe();
+return conditionResultUnwrapped;
+```
+3. Add the constraint that `val` has type `A` where there's a function called `isSome` with the signature `(A) -> X`, with `X` constrained to a `bool` due to its use in the conditional part of an `if` statement.
+4. Add the constraint that `val` has type `B` where there's a function called `getMaybe` with the signature `(B) -> C`.
+5. Add the constraint that type `C` from `getMaybe` has a function called `toString` with the signature `(C) -> U`.
+6. Add the constraint that type `U` is the same as the concrete type `string` (this is from the `return 'none'` statement).
+7. Reduce type `A` to only have the possibility of being type `Maybe<S>`.
+8. Reduce type `B` to only have the possibility of being type `Maybe<R>`.
+9. Since `Stringifiable` is an already-existing interface that perfectly matches the constraints on type `C`, replace all instances of type `C` with `Stringifiable`.
+10. Since `A` and `B` both are in reference to the same variable, add the constraint that they satisfy each other.
+11. Reduce type `Maybe<R>` to be equivalent to the type `Maybe<Stringifiable>`
+12. `Maybe<Stringifiable>` is the minimal interface value for `val`, so the signature of the `toString` fn is now rewritten as `(Maybe<Stringifiable>) -> U`.
+13. `string` is the concrete type representation for `U`, so the signature of the `toString` fn is now rewritten as `(Maybe<Stringifiable>) -> string`.
+
+(side note: the `getMaybe()` call will also eventually be syntactic sugar, according to [RFC 12][RFC-12]).
+
+### Alternatives Considered
+
+One alternative is to keep the current type system and compiler front-end and adding changes as-needed.
+However, the current front-end's design means that it takes relatively substantial work to add each feature, with [new features potentially exposing "load-bearing" bugs][cond-table-PR].
+Additionally, there is [a lot of documented tech-debt][tech-debt-issues] that would already result in substantial parts of the compiler being reworked or rewritten.
+Lastly, the Hindley-Milner type system has been well-documented and proven both [over time][haskell-type-system-revered] and [in theory][emlti].
```

Comment:
I'm not familiar with them, but are there other type systems that should be considered here?

Reply:
There are alternatives, but only the HM scheme has been proven to not require type annotations anywhere in order to produce a valid program.
- C++/Java-style type inference is rather weak and fails in many cases - they're unable to adapt more complex algorithms due to subtyping in their system (class extensions)
  - Dart has subtyping as well and its type inference is more extensive, but [still has restrictions](https://github.com/dart-lang/sdk/issues/37247)
  - Swift has [similar issues](https://cs.stackexchange.com/questions/26389/why-isnt-the-swift-programming-language-type-inference-more-aggressive)
- Rust is [a modified version of Hindley-Milner](https://rustc-dev-guide.rust-lang.org/type-inference.html)
- Kotlin also uses [a partial version Hindley-Milner](https://kotlinlang.org/spec/type-inference.html), but again subtyping means that type annotations are needed in certain parts of the language
- TypeScript uses a [candidate system similar to the current front-end, but requires multiple passes to correctly determine the type of an object](https://github.com/microsoft/TypeScript/issues/2182) - and even then, it can't determine the concrete types of values all the time, [defaulting to `any` in many scenarios where it can't determine the type of an object](https://basarat.gitbook.io/typescript/type-system/type-inference)
- Golang [uses the C/C++/Java-style type inference (despite not having sub-typing)](https://stackoverflow.com/questions/12589255/gos-type-inference-algorithm), but its otherwise simple syntax and obligatory type declarations in function syntax or for interfaces/structs, etc means that it's still quite effective
- Typed Racket [is Hindley-Milner](https://docs.racket-lang.org/plai-typed/Type_Checking_and_Inference.html)